File: Emulator/src/main/content/global/skill/agility/AgilityCourse.java
package content.global.skill.agility;

import core.game.interaction.OptionHandler;
import core.game.node.entity.player.Player;
import core.game.node.entity.skill.Skills;
import core.plugin.Plugin;

/**
 * An abstract base class representing an agility course consisting of multiple obstacles.
 *
 * @author Emperor
 */
public abstract class AgilityCourse extends OptionHandler {
    private final Player player;
    private final boolean[] obstaclesPassed;
    private final double completionExperience;

    /**
     * Constructs a new {@code AgilityCourse}.
     *
     * @param player               the player completing the course
     * @param size                 the number of obstacles in the course
     * @param completionExperience the amount of Agility experience awarded upon completion
     */
    public AgilityCourse(Player player, int size, double completionExperience) {
        this.player = player;
        this.obstaclesPassed = new boolean[size];
        this.completionExperience = completionExperience;
    }

    @Override
    public final Plugin<Object> newInstance(Object arg) throws Throwable {
        if (arg == null) {
            configure();
            return this;
        }
        return createInstance((Player) arg);
    }

    @Override
    public Object fireEvent(String identifier, Object... args) {
        return null;
    }

    /**
     * Performs any course-specific configuration, such as registering obstacle interactions.
     */
    public abstract void configure();

    /**
     * Creates a new instance of the agility course for the given player.
     *
     * @param player the player
     * @return a new instance of the course
     */
    public abstract AgilityCourse createInstance(Player player);

    /**
     * Gets the active course instance for the specified player, creating one if necessary.
     *
     * @param player the player
     * @return the agility course instance
     */
    public AgilityCourse getCourse(Player player) {
        AgilityCourse course = player.getExtension(AgilityCourse.class);
        if (course == null || course.getClass() != getClass()) {
            try {
                player.addExtension(AgilityCourse.class, course = createInstance(player));
            } catch (Throwable e) {
                e.printStackTrace();
            }
        }
        return course;
    }

    /**
     * Marks the course as completed if all obstacles were passed and awards experience.
     */
    public void finish() {
        if (isCompleted()) {
            player.getSkills().addExperience(Skills.AGILITY, completionExperience, true);
        }
        reset();
    }

    /**
     * Checks whether all obstacles in the course have been successfully completed.
     *
     * @return {@code true} if the course is complete, {@code false} otherwise
     */
    public boolean isCompleted() {
        for (boolean b : obstaclesPassed) {
            if (!b) {
                return false;
            }
        }
        return true;
    }

    /**
     * Resets all obstacle progress for this course.
     */
    public void reset() {
        for (int i = 0; i < obstaclesPassed.length; i++) {
            obstaclesPassed[i] = false;
        }
    }

    /**
     * Flags a specific obstacle as completed. If it's the final obstacle, the course is finished.
     *
     * @param index the index of the completed obstacle
     */
    public void flag(int index) {
        obstaclesPassed[index] = true;
        if (index == obstaclesPassed.length - 1) {
            finish();
        }
    }

    /**
     * Calculates the damage to take when failing an obstacle, based on the player's health.
     *
     * @param player the player
     * @return the damage amount
     */
    protected static int getHitAmount(Player player) {
        int hit = player.getSkills().getLifepoints() / 12;
        return Math.max(hit, 2);
    }

    /**
     * Returns the player associated with this course.
     *
     * @return the player
     */
    public Player getPlayer() {
        return player;
    }

    /**
     * Returns the array indicating which obstacles have been passed.
     *
     * @return a boolean array of passed obstacle flags
     */
    public boolean[] getPassedObstacles() {
        return obstaclesPassed;
    }
}

File: Emulator/src/main/content/global/skill/agility/AgilityCourse.kt
package content.global.skill.agility

import core.game.interaction.OptionHandler
import core.game.node.entity.player.Player
import core.game.node.entity.skill.Skills
import core.plugin.Plugin

/**
 * An abstract base class representing an agility course consisting of multiple obstacles.
 *
 * @author Emperor
 */
abstract class AgilityCourse(
    private val player: Player,
    size: Int,
    private val completionExperience: Double
) : OptionHandler() {

    private val obstaclesPassed = BooleanArray(size)

    @Throws(Throwable::class)
    override fun newInstance(arg: Any?): Plugin<Any> {
        return if (arg == null) {
            configure()
            this
        } else {
            createInstance(arg as Player)
        }
    }

    override fun fireEvent(identifier: String, vararg args: Any?): Any? {
        return null
    }

    /**
     * Performs any course-specific configuration, such as registering obstacle interactions.
     */
    abstract fun configure()

    /**
     * Creates a new instance of the agility course for the given player.
     *
     * @param player the player
     * @return a new instance of the course
     */
    abstract fun createInstance(player: Player): AgilityCourse

    /**
     * Gets the active course instance for the specified player, creating one if necessary.
     *
     * @param player the player
     * @return the agility course instance
     */
    fun getCourse(player: Player): AgilityCourse {
        var course = player.getExtension<AgilityCourse>(AgilityCourse::class.java)
        if (course == null || course::class != this::class) {
            try {
                course = createInstance(player)
                player.addExtension(AgilityCourse::class.java, course)
            } catch (e: Throwable) {
                e.printStackTrace()
            }
        }
        return course
    }

    /**
     * Marks the course as completed if all obstacles were passed and awards experience.
     */
    fun finish() {
        if (isCompleted()) {
            player.skills.addExperience(Skills.AGILITY, completionExperience, true)
        }
        reset()
    }

    /**
     * Checks whether all obstacles in the course have been successfully completed.
     *
     * @return true if the course is complete, false otherwise
     */
    fun isCompleted(): Boolean = obstaclesPassed.all { it }

    /**
     * Resets all obstacle progress for this course.
     */
    fun reset() {
        obstaclesPassed.fill(false)
    }

    /**
     * Flags a specific obstacle as completed. If it's the final obstacle, the course is finished.
     *
     * @param index the index of the completed obstacle
     */
    fun flag(index: Int) {
        obstaclesPassed[index] = true
        if (index == obstaclesPassed.lastIndex) {
            finish()
        }
    }

    companion object {
        /**
         * Calculates the damage to take when failing an obstacle, based on the player's health.
         *
         * @param player the player
         * @return the damage amount
         */
        fun getHitAmount(player: Player): Int {
            val hit = player.skills.lifepoints / 12
            return maxOf(hit, 2)
        }
    }

    /**
     * Returns the player associated with this course.
     */
    fun getPlayer(): Player = player

    /**
     * Returns the array indicating which obstacles have been passed.
     */
    fun getPassedObstacles(): BooleanArray = obstaclesPassed
}
